pkg/output/stream_test.go

```diff
@@ -0,0 +1,168 @@
+)
+
+// Hook up gocheck into the "go test" runner.
+func Test(t *testing.T) { TestingT(t) }
+
+type OutputTestSuite struct{}
+
+var _ = Suite(&OutputTestSuite{})
+
+type testCase struct {
+	prefixLength      int
+	prefixWithEndline bool
+	key               string
+	value             []rune
+}
+
+func generateLength(r *rand.Rand, avgLength int) int {
+	if avgLength == 0 {
+		return 0
+	}
+	return r.Intn(avgLength/5) + avgLength // Return random length ±20% of avgLength
+}
+
+var runes = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_!=-+*/\\")
+
+func generateRandomRunes(r *rand.Rand, length int, endline rune) []rune {
+	totalLength := length
+	if endline != NoEndline {
+		totalLength += 1
+	}
+	line := make([]rune, totalLength)
+	var last rune
+	for j := 0; j < length; j++ {
+		var current rune
+		for rpt := true; rpt; rpt = last == '\\' && (current == '\n' || current == '\r') {
+			current = runes[r.Intn(len(runes))]
+		}
+
+		line[j] = current
+		last = current
+	}
+
+	if endline != NoEndline {
+		line[length] = endline
+	}
+	return line
+}
+
+func generateTestCases(numOfLines, avgPrefixLength, avgKeyLength, avgValueLength int, endlinePolicy EndlinePolicy) []testCase {
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+	cases := make([]testCase, numOfLines)
+	for i := 0; i < numOfLines; i++ {
+		key := ""
+		value := []rune{}
+		if avgKeyLength != 0 {
+			key = apirand.String(generateLength(r, avgKeyLength))
+			value = generateRandomRunes(r, avgValueLength, NoEndline)
+		}
+		cases[i] = testCase{
+			prefixLength:      generateLength(r, avgPrefixLength),
+			prefixWithEndline: endlinePolicy == EndlineRequired || rand.Intn(2) == 1,
+			key:               key,
+			value:             value,
+		}
+	}
+
+	return cases
+}
+
+func getTestReaderCloser(done chan struct{}, cases []testCase) io.ReadCloser {
+	pr, pw := io.Pipe()
+	r := rand.New(rand.NewSource(time.Now().UnixNano()))
+
+	go func() {
+		defer pw.Close()
+
+		for _, tc := range cases {
+			select {
+			case <-done:
+				return
+			default:
+				if tc.prefixLength != 0 {
+					endline := NoEndline
+					if tc.prefixWithEndline {
+						endline = NewlineEndline
+					}
+					prefixLine := generateRandomRunes(r, tc.prefixLength, endline)
+					_, err := pw.Write([]byte(string(prefixLine)))
+					if err != nil {
+						return
+					}
+				}
+
+				if tc.key != "" {
+					err := output.PrintOutputTo(pw, tc.key, string(tc.value))
+					if err != nil {
+						return
+					}
+				}
+			}
+		}
+	}()
+
+	return pr
+}
+
+func (s *OutputTestSuite) TestHugeStreamsWithoutPhaseOutput(c *C) {
+	done := make(chan struct{})
+	defer func() { close(done) }()
+
+	// e-sumin: Here I'm generating test case, when we have just random string around 10000 runes
+	// I expect that it will be logged as one line as it was before
+	// But in fact it is logged by chunks of ±4kb
+	// When we will fix the code behavior, numOfLines has to be set to 10, and avgPrefix len has to be set to 500000
```

Comment:
@hairyhum could you please check this case ? I believe we should try to keep old behaviour as soon as it possible.

Reply:
Fixed with replacing `NewReader` with `NewReaderSize`